### ISSUE TITLE
6-Dimensional Ferret

### ISSUE THREAD

**Issue by karlmsmith** [NOAA-PMEL/Ferret#815]
**Reported by Ansley B Manke on 22 Oct 2007 19:25 UTC**
Collecting some past email about the topic of extending Ferret to 6 dimensions:

6-dimensional Ferret 22-Oct-2007 Steve to Jing and Ansley
Hi Jing,

I think we may have talked about this topic a bit in the past:  the need to
expand Ferret from a 4-dimensional environment to a 6-dimensional environment. 
Ansley has given her OK to working on this team-fashion and I hope you'll agree
to be a partner in that team.  This email is to begin to evolve that discussion
into a work task.  And to discuss how to prioritize it against other tasks that
you are working on.  This is a big enough task that I imagine you will want to
take a "4 hours per day" approach to it initially -- work on it in parallel with
other tasks.  Please let me know your initial thoughts about this.

    - Steve

=====

The requirements that motivate this are the need to support 1) a forecast (time)
axis and 2) an ensemble (index) axis.  There will be basic questions about
Ferret's command syntax to sort out.  For now lets refer to the new axes in
world coords as E and F -- X,Y,Z,T,E,F -- and in index coordinates as
i,j,k,l,m,n.  But lets be alert for better name choices in the future.  These
choices would break some existing scripts, because these names are commonly used
as variable names.

How to convert the 4D logic in Ferret into 6D logic?  Given the limitations of
FORTRAN syntax this probably has to be a brute force exercise in source code
editing.  The big question will be how to accomplish it at minimum effort.  For
example, could you envision a perl script (or sed if you are supremely skilled)
that could look for the syntax  *(*,*,*,*)  and make a good guess at the
corresponding *(*,*,*,*,*,*) in order to get much of the brute force editing
done automatically?  Maybe or maybe not.  (It might make the problem bigger
instead of smaller.)  Such a script would need to be sensitive to FORTRAN
continuation lines and line length.  And often the (*,*,*,*) pattern treats one
axis specially, so perl code would need to look for the "*" pattern that was
followed at least 3 times and use that as the basis for adding the two new
dimensions.  You'd probably want to refine and refine and refine the perl code
-- with test compilations between each iteration -- until as much as possible of
the conversion was automated before starting any manual edits.

We'd also want to try to anticipate the places where there will be serious
additions to the code *logic*:  E.g.

    * netCDF IO ("tmap library" code in the tmf directory)
    * external function support utilities
    * low level utilities, like cx_dim that compute the size or dimensionality
of an "object"
    * E,F,m,n as pseudovariables
          o "set mode 4-dimensional" to ignore E,F,m,n

My gut instinct is that the enhancement should be done in three stages: the
first stage merely confirms that the range of compilers we work with can handle
6-dimensional variables "var(ilo:ihi, jlo:jhi, klo:khi, llo:lhi, mlo:mhi,
nlo:nhi,)".  The second stage reproduces all of the current behaviors (4D), but
with source code variables that are internally declared as 6D (the upper 2
dimensions never being referenced except as -111:-111).  The third stage would
be the changes to the code logic.  This enhancement may also be the right
opportunity to add some level of curvilinear grid "intelligence" into Ferret --
making Ferret understand that some variables are defined on curvilinear grids.

Re: Three topics -- 6 dimensional Ferret 3/7/2007 Steve to Ansley

 Ferret 6D:
It is very likely that we will need to support 6-dimensional netCDF files in the
future.  The two additional axes could be anything, bu their designated roles
will be 1) a forecast (time) axis and 2) and ensemble (index) axis.  There will
be all sorts of questions about Ferret's command syntax to sort out.  For this
discussion I'll refer to the axes in world coords as X,Y,Z,T,E,F and in index
coordinates as i,j,k,l,m,n.  (Of course these choices would break some existing
scripts, because of pseudo-variable names clobbering variable names.)

The big question will be how to implement this.  It is not too soon to begin
thinking about this, as it makes sense to mull the ideas over for a while.  (It
may also be the right opportunity to add some level of curvilinear grid
"intelligence" into Ferret -- making Ferret understand that some variables are
defined on curvilinear grids.)

The brute force approach is simply to convert all of the 4D logic into 6D logic.
 Maybe this is the best way to go.  Huge amounts of code editing (see discussion
below), but at the end an uncompromising full implementation of two new dimensions.

The other approach is to retain all of the 4D logic internally, permitting 6D
variables to exist only as file variables.  In other words Ferret would support
reading at most 4D subsets from a file variable at any given time.  The
challenge to this is that every expression in Ferret has to have a grid
determined at the time that the expression is defined -- prior to evaluation. 
This means that the restriction of 6D vars down to 4D must be either implicit or
explicit when the expression is first introduced.  For example, if "V6" is a 6D
variable then
    v6[a valid expression.  As is
    v6[X=180,Z=0](p=1,q=1]
is)

Please give some thought to the two approaches -- particularly to envision the
effort to do the brute force approach, since at the end it doesn't complicate
the code logic so much.  I wonder, for example, if you could envision a perl
script (or sed if you are supremely skilled) that could look for the syntax
   *(*,*,*,*)
and make a good guess at the corresponding *(*,*,*,*,*,*) in order to get much
of the brute force editing done automatically.  It would need to be sensitive to
FORTRAN continuation lines and line length.  And often the (*,*,*,*) pattern
treats one axis specially, so your perl code would need to look for the "*"
pattern that was followed at least 3 times and use that as the basis for adding
the two new dimensions.  You'd want to refine and refine and refine the perl
code -- with test compilations between each iteration -- until as much as
possible of the conversion was automated before starting any manual edits.

You could imagine the full 6D (brute force) being done in two stages, where the
first stage merely reproduced all of the current behaviors (4D), but using
variables that are internally declared as 6D (the upper 2 dimensions never being
referenced except as -999:-999).
A lot to think about ... 

Re: Three topics -- 6 dimensional Ferret 3/14/2007 from Ansley to Steve
...
We'd need to be careful about string variables where an extra dimension is
already in those bits of code, keeping track of the string length. 
...

Migrated-From: http://dunkel.pmel.noaa.gov/trac/ferret/ticket/1543

**Comment by karlmsmith** [comment]
**Comment by ansley.b.manke on 13 Feb 2012 17:37 UTC**
Move this comment and the following one to a new ticket, #1917.

**Comment by karlmsmith** [comment]
**Modified by steven.c.hankin on 13 Feb 2012 18:35 UTC**

**Comment by karlmsmith** [comment]
**Comment by ansley.b.manke on 24 Feb 2012 00:41 UTC**
Bumping up the priority of this so it's easier to find...

**Comment by karlmsmith** [comment]
**Comment by ansley.b.manke on 24 Feb 2012 00:47 UTC**
And adding Steve's proposed breakdown of work (email of 2/14/2012)

I propose this breakdown of the effort:

    Karl - handle the code changes under the fer directory, except for external functions efi, efn -- only through Stage 2, below.  I anticipate there to be around 400 routines requiring changes.  The overwhelming majority of changes will be very routine in nature ... mostly just adding 2 additional arguments in the parentheses, changing dimensions from 4 to 6 and occasionally adding 2 more nested loops surrounding the current loops.   At this stage commands, like SHOW DATA can simply continue doing 4D displays.  
        Probably add in some LOGICAL variables to indicate when 6D can be ignored:  LOGICAL ds_is_4d
    Ansley - handle the changes to the TMAP_FORMAT libraries (IO), the external functions, and most of the significant *behavior* changes.  EFs will need to have a 4D legacy mode.

The idea of this approach, Karl, is to limit the time that you will need to be committed to this.  Just 2-3 weeks of help with text editing, then you should be free of this.  I suggest that you stay in tight coordination initially, when upgrading the COMMON files.

I propose that the implementation be conducted in three stages:

    Ansley (1 day):  confirm that the range of compilers we work with can handle 6-dimensional variables "var(ilo:ihi, jlo:jhi, klo:khi, llo:lhi, mlo:mhi, nlo:nhi,)".  Just write a short test program and confirm that it works.  We'll all look it over together to make sure it tests any concerns that we may have.
    Both (3 weeks):  reproduce all of the current Ferret behaviors (4D), but with source code variables that are internally declared as 6D (the upper 2 dimensions never being referenced except as -111:-111).   Not all changes need have been considered at this stage.   For example, introducing the new pseudo-variables,  "E" and "F", can wait until stage 3.
    Mostly Ansley:  changes to the code logic so that the 5th and 6th dimensions function correctly -- "E" and "F", LET variables, EFs, IO, SHOW DATA

**Comment by karlmsmith** [comment]
**Modified by ansley.b.manke on 4 Mar 2013 18:58 UTC**
